x-pack/plugins/security_solution/server/endpoint/routes/metadata/support/unenroll.ts

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { APICaller, IScopedClusterClient } from 'kibana/server';
+import { SearchResponse } from 'elasticsearch';
+import { metadataMirrorIndexPattern } from '../../../../../common/endpoint/constants';
+import { EndpointStatus } from '../../../../../common/endpoint/types';
+
+const KEEPALIVE = '10s';
+
+export interface HostId {
+  host: {
+    id: string;
+  };
+}
+
+interface HitSource {
+  _source: HostId;
+}
+
+export async function findUnenrolledHostByHostId(
+  client: IScopedClusterClient,
+  hostId: string
+): Promise<HostId | undefined> {
+  const queryParams = {
+    index: metadataMirrorIndexPattern,
+    body: {
+      size: 1,
+      _source: ['host.id'],
+      query: {
+        bool: {
+          must: {
+            term: { 'host.id': hostId },
+          },
+          filter: {
+            term: {
+              'Endpoint.status': EndpointStatus.UNENROLLED,
+            },
+          },
+        },
+      },
+    },
+  };
+
+  const response = (await client.callAsCurrentUser('search', queryParams)) as SearchResponse<
+    HostId
+  >;
+  const newHits = response.hits?.hits || [];
+
+  if (newHits.length > 0) {
+    const hostIds = newHits
+      .flatMap((data) => data as HitSource)
+      .map((hitSource: HitSource) => hitSource._source);
+    return hostIds[0];
+  } else {
+    return undefined;
+  }
+}
+
+export async function findAllUnenrolledHostIds(client: IScopedClusterClient): Promise<HostId[]> {
+  const queryParams = {
+    index: metadataMirrorIndexPattern,
+    scroll: KEEPALIVE,
+    body: {
+      size: 100,
+      _source: ['host.id'],
+      query: {
+        bool: {
+          filter: {
+            term: {
+              'Endpoint.status': EndpointStatus.UNENROLLED,
+            },
+          },
+        },
+      },
+    },
+  };
+  const response = (await client.callAsCurrentUser('search', queryParams)) as SearchResponse<
+    HostId
+  >;
+
+  // eslint-disable-next-line no-return-await
+  return await fetchAllUnenrolledHostIdsWithScroll(response, client.callAsCurrentUser);
+}
+
+export async function fetchAllUnenrolledHostIdsWithScroll(
+  response: SearchResponse<HostId>,
+  client: APICaller,
+  hits: HostId[] = []
+): Promise<HostId[]> {
+  const newHits = response.hits?.hits || [];
+  const scrollId = response._scroll_id;
+
+  if (newHits.length > 0) {
+    const hostIds: HostId[] = newHits
+      .flatMap((data) => data as HitSource)
+      .map((hitSource: HitSource) => hitSource._source);
+    hits.push(...hostIds);
+
+    const innerResponse = await client('scroll', {
```

Comment:
Hmm, I don't think we want to use `scroll` here. The docs say:

`Scrolling is not intended for real time user requests, but rather for processing large amounts of data, e.g. in order to reindex the contents of one index into a new index with a different configuration.`

https://www.elastic.co/guide/en/elasticsearch/reference/7.8/search-request-body.html#request-body-search-scroll

I think we want to use `search_after` instead: https://www.elastic.co/guide/en/elasticsearch/reference/7.8/search-request-body.html#request-body-search-search-after

Reply:
We discussed this offline. The function here is to retrieve all the results (only the ids) using a cursor based approach and does not involve any user interaction or pagination. This approach was taken in the event that the number of unerolled endpoints are greater than 10000, which is the default max size of a page. We are also returning only the host ids, so we are very minimal. Only returning the IDs means faster transmission and processing time, and small space requirement . We also keep the previous search context for 10s to 30s. I increased to 30s with the increase in page size, to allow for processing.